pydocstringformatter/_formatting/base.py

```diff
@@ -202,3 +206,91 @@ def treat_summary(
 
     def treat_description(self, description: str, indent_length: int) -> str:
         return description
+
+
+def _pairwise(iterator: Iterator[_T]) -> Iterator[tuple[_T, _T]]:
+    """Create an iterator over pairs of successive elements."""
+    first, second = tee(iterator)
+    next(second)
+    return zip(first, second)
+
+
+class NumpydocSectionFormatter(StringAndQuotesFormatter, metaclass=abc.ABCMeta):
+    """Base class for formatters working on numpydoc sections."""
+
+    style = ["numpydoc"]
+
+    @abc.abstractmethod
+    def treat_sections(
+        self, sections: OrderedDict[str, list[str]]
+    ) -> OrderedDict[str, list[str]]:
+        """Process the individual numpydoc sections."""
+
+    def treat_string(
+        self,
+        tokeninfo: tokenize.TokenInfo,
+        indent_length: int,
+        quotes: str,
+        quotes_length: Literal[1, 3],
+    ) -> str:
+        """Split numpydoc sections, pass them for processing, then rejoin them."""
+        # Split sections
+        lines = tokeninfo.string[quotes_length:-quotes_length].split("\n")
+        # handle the spaces before the closing quotes
+        last_line = lines[-1]
+        if lines[-1].isspace():
+            lines[-1] = ""
+
+        # Everything before the first section header is in a single
+        # summary/deprecation warning/extended summary section.  This
+        # ends up called "Summary".
+        section_hyphen_lines = [
+            index
+            for index, line in enumerate(lines)
+            if "-" in line and all(char in " \t-" for char in line)
+        ]
+        section_starts = (
+            [0] + [index - 1 for index in section_hyphen_lines] + [len(lines)]
+        )
+
+        sections = OrderedDict(
+            [
+                (
+                    lines[curr_section_start].lstrip(),
+                    lines[curr_section_start:next_section_start],
+                )
+                for curr_section_start, next_section_start in _pairwise(
+                    iter(section_starts)
+                )
+            ]
+        )
+        if section_hyphen_lines and section_hyphen_lines[0] == 1:
+            # No summary/deprecation warning/extended summary section
+            section_starts.pop(0)
```

Comment:
Can `section_starts` be a `set` to avoid this?

Reply:
I need that list sorted and I'm not sure that `set`s are ordered the way `dict`s have been since 3.7 or so.  This is the last time that variable is used, and dictionary semantics means most other problems are already dealt with; would you prefer I negated the condition and only included the other branch?

Reply:
Yeah, let's only do the other branch! I didn't see it wasn't used anymore.